### DISCUSSION TITLE
Transaction events (beforeTransactionCommit, afterTransactionCommit ...) not triggered on nested transactions

### DISCUSSION THREAD

**Discussion by nicomouss** [mikro-orm/mikro-orm#5033] [General]
### Describe the bug

Not sure this is a bug, or the intended behaviour so far.
So this might indeed fall into an enhancement. Sorry if haven't targeted the right issue type.

Let me first state that mikro-orm works beautifully regarding nesting transactional contexts, leading to very elegant code. In our case, it enables to implement individual services operations as atomic executions: each operation is wrapped into its own transactional context, and can be individually executed. Later on, multiple services operations can be grouped into a single atomic unit by wrapping them into a "master" service (master transactional context). Using decorators on services' methods, everything is transparent and handled under the hood by mikro-orm. Awesome stuff.

Now, we try to use event subscribers to track down the transactions events of each of our transactional contexts, nested or not.

Transaction events work fine when no transactional context are nested.
However, when nesting transactional contexts, mikro-orm only emits transaction events for the master transactional context, and not for the nested ones.

Please check the example code below to reproduce the behaviour.

### Reproduction

```ts
await orm.em.transactional(
    async (transactionalEntityManager: SqlEntityManager) => {
        /* Tracking events for this master transactional context ... */
        transactionalEntityManager.getEventManager().registerSubscriber({
            beforeTransactionCommit: async (args) => {
                console.log("### master transaction - beforeTransactionCommit");
            },
            afterTransactionCommit: (args) => {
                console.log("### master transaction - afterTransactionCommit");
            },
        });

        await transactionalEntityManager.transactional(
            async (nestedTransactionalEntityManager1: SqlEntityManager) => {
                /* Tracking events for this nested transactional context ... */
                nestedTransactionalEntityManager1.getEventManager().registerSubscriber({
                    beforeTransactionCommit: async (args) => {
                        console.log("### nested transaction 1 - beforeTransactionCommit");
                    },
                    afterTransactionCommit: (args) => {
                        console.log("### nested transaction 1 - afterTransactionCommit");
                    },
                });

                /* business logic here ... */
            },
        );

        await transactionalEntityManager.transactional(
            async (nestedTransactionalEntityManager2: SqlEntityManager) => {
                /* Tracking events for this nested transactional context ... */
                nestedTransactionalEntityManager2.getEventManager().registerSubscriber({
                    beforeTransactionCommit: async (args) => {
                        console.log("### nested transaction 2 - beforeTransactionCommit");
                    },
                    afterTransactionCommit: (args) => {
                        console.log("### nested transaction 2 - afterTransactionCommit");
                    },
                });

                /* business logic here ... */
            },
        );
    },
);
```

This produces the following output:
```bash
### master transaction - beforeTransactionCommit
### master transaction - afterTransactionCommit
```

Would expect it to produce something like:
```bash
### nested transaction 1 - beforeTransactionCommit
### nested transaction 1 - afterTransactionCommit
### nested transaction 2 - beforeTransactionCommit
### nested transaction 2- afterTransactionCommit
### master transaction - beforeTransactionCommit
### master transaction - afterTransactionCommit
```

I understand that nested transactions are implemented at the database level using savepoints. So technically speaking, only the master transaction is in effect triggering real ```BEGIN``` and ```COMMIT``` statements.

But at the app level, the nested transactions are also considered as proper transactions, so they should be logically treated as so, and raise proper events (in my opinion). This would mean that there should be events raised whenever a savepoint is released (meaning the nested transaction is committed at the database level) or rollbacked (meaning the nested transaction is rollbacked at the database level).

In the subscribers, there could be for example a boolean in the event args, indicating if the transaction is a nested one (or is the master transaction):

```ts
transactionalEntityManager.getEventManager().registerSubscriber({
    beforeTransactionCommit: async (args) => {
        if (args.nestedTransaction) 
           console.log("### the transaction is a nested one, and is about to commit !");
    }
});
```

Or there could be different event raised, depending if the related transaction is nested or not:

```ts
transactionalEntityManager.getEventManager().registerSubscriber({
    beforeTransactionCommit: async (args) => {
        console.log("### the transaction is a master one, and is about to commit !");
    },
    beforeNestedTransactionCommit: async (args) => {
        console.log("### the transaction is nested one, and is about to commit !");
    }
});
```

The second solution would bring no breaking change to existing code bases that use mikro-orm.

Hope those explanations makes sense.
happy to discuss this further.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

5.9.4

### Node.js version

18

### Operating system

MacOs

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
They are indeed fired only for the root transactions. I'd say it's because the nested transactions don't really commit, they use savepoints instead. 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/AbstractSqlConnection.ts#L119

I don't have strong opinions about this and I don't see any other reason why those hooks shouldn't be fired, so will be fine with removing this check from v6.

@rhyek since you've contributed this (#1213), what's your opinion? Do you by any chance remember why the check is there?

**Reply by nicomouss** [reply]
> I don't have strong opinions about this and I don't see any other reason why those hooks shouldn't be fired, so will be fine with removing this check from v6.

@B4nan @rhyek I guess the check is still there in the code in v6+, is there anything new around this ?

Maybe an idea would be to add a config option. Meaning something like `enableNestedTransactionEvents`, which would default to `false` so that mikroorm would behave by default the same as it does today. Logic could then be:

If `enableNestedTransactionEvents` is set to false: transaction events are only raised for the root transaction.
If `enableNestedTransactionEvents` is set to true: transaction events are raised for both the root transaction and the nested ones.

**Reply by B4nan** [reply]
I think I will just do this in v7 without any opt-out, users can ignore the nested events based on some parameter. 

If you need this in v6, I'd accept a PR with the opt-in variant too. But it will result in some nasty merge conflicts with the v7 branch likely 🙃

**Reply by nicomouss** [reply]
Sure, all good if those events are fired without any opt-out in v7 !
Was only proposing this opt-in idea in case the change would add some internal side-effects I might not been aware to.

Then that sounds perfectly legit to ignore the nested events based on some parameters, if needed. Makes actually things simpler by not adding some "config overhead".

Thanks for your reply.

**Reply by B4nan** [reply]
7b73ebfd1bbcfd0b264b736b33c3f049c2ec6772

**Reply by nicomouss** [reply]
Wonderful !

Just to be sure, isn't there a typo in the committed documentation ?

It is written `All transaction events are not invoked for child transactions, too.`
Shouldn't it be `All transaction events are invoked for child transactions, too.` ?

**Reply by B4nan** [reply]
Right 8926a3ec7958f4d4a303c549c08d25a039b9c2d8
